### PR TITLE
LPS-90257 Add dropdown menu to the User Personal Bar

### PR DESCRIPTION
**Pull request by pei-jung**
Hi Chema,

I've made the changes per your comments in https://github.com/jbalsas/liferay-portal/pull/1678. Everything seems to work fine except the menu still shows when a menu item is clicked and navigated to a different page. I tried debugging it, but it seems that no `disposed` nor `detached` is fired when navigating to a different page. Can you take a look? Thanks!

### CODE REVIEW

File: modules/apps/product-navigation/product-navigation-user-personal-bar-web/src/main/resources/META-INF/resources/view.jsp [hunk 1]
```diff
@@ -73,20 +73,38 @@ portletURL.setWindowState(LiferayWindowState.EXCLUSIVE);
...
+					credentials: 'include',
+					headers: {
+						'Content-Type': 'html'
+					},
+					method: 'GET'
 				}
-			})
+			).then(
+				function(response) {
+					return response.text();
```
Comment (jbalsas) [inline modules/apps/product-navigation/product-navigation-user-personal-bar-web/src/main/resources/META-INF/resources/view.jsp]: Shouldn't this be simple json? What's this actually doing?

File: modules/apps/product-navigation/product-navigation-user-personal-bar-web/src/main/resources/META-INF/resources/view.jsp [hunk 2]
```diff
@@ -73,20 +73,38 @@ portletURL.setWindowState(LiferayWindowState.EXCLUSIVE);
...
+				}
+			).then(
+				function(data) {
+					userAvatar.innerHTML = data;
+
+					var scripts = userAvatar.getElementsByTagName("script");
+
+					for (var i = 0; i < scripts.length; i++) {
+						var script = scripts[i];
+						eval(script.innerHTML);
```
Comment (jbalsas) [inline modules/apps/product-navigation/product-navigation-user-personal-bar-web/src/main/resources/META-INF/resources/view.jsp]: Using `eval` is **evil** 😉 

Please, search for `dom.globalEval.runScriptsInElement` for an alternative way to do this!

File: modules/apps/product-navigation/product-navigation-user-personal-bar-web/src/main/resources/META-INF/resources/view.jsp [hunk 3]
```diff
@@ -73,20 +73,38 @@ portletURL.setWindowState(LiferayWindowState.EXCLUSIVE);
...
+					method: 'GET'
 				}
-			})
+			).then(
+				function(response) {
+					return response.text();
+				}
+			).then(
+				function(data) {
+					userAvatar.innerHTML = data;
```
Comment (jbalsas) [inline modules/apps/product-navigation/product-navigation-user-personal-bar-web/src/main/resources/META-INF/resources/view.jsp]: Can you explain what is this doing? why are we getting back html, and JS if we had this in the first place?

File: modules/apps/product-navigation/product-navigation-taglib/src/main/resources/META-INF/resources/user_personal_menu/page.jsp [hunk 4]
```diff
@@ -18,13 +18,47 @@
...
+						if (<%= expanded %>) {
+							this.expanded = true;
+						}
+
+						var menu = this.refs.dropdown.refs.portal.refs.menu;
+
+						menu.style.maxHeight = 'fit-content';
+					}
+				},
+				items: <%= dropdownItems.toJSONString() %>,
```
Comment (jbalsas) [inline modules/apps/product-navigation/product-navigation-taglib/src/main/resources/META-INF/resources/user_personal_menu/page.jsp]: We already have the items... so we shouldn't need to do anything else, do we?

File: modules/apps/product-navigation/product-navigation-user-personal-bar-web/src/main/resources/META-INF/resources/view.jsp (hunk 1 above)
Comment (pei-jung) [reply modules/apps/product-navigation/product-navigation-user-personal-bar-web/src/main/resources/META-INF/resources/view.jsp]: We want to load the dropdown menu only when the user avatar is clicked, so I'm doing an ajax on click to get the dropdown menu.

File: modules/apps/product-navigation/product-navigation-user-personal-bar-web/src/main/resources/META-INF/resources/view.jsp (hunk 2 above)
Comment (pei-jung) [reply modules/apps/product-navigation/product-navigation-user-personal-bar-web/src/main/resources/META-INF/resources/view.jsp]: Ah, good to know. I'll look into that.

File: modules/apps/product-navigation/product-navigation-user-personal-bar-web/src/main/resources/META-INF/resources/view.jsp (hunk 3 above)
Comment (pei-jung) [reply modules/apps/product-navigation/product-navigation-user-personal-bar-web/src/main/resources/META-INF/resources/view.jsp]: The userAvatar is just an image with text. I'm replacing that with the dropdown menu here.

File: modules/apps/product-navigation/product-navigation-taglib/src/main/resources/META-INF/resources/user_personal_menu/page.jsp (hunk 4 above)
Comment (pei-jung) [reply modules/apps/product-navigation/product-navigation-taglib/src/main/resources/META-INF/resources/user_personal_menu/page.jsp]: Are you talking about the custom event listener that I added? That's for automatically expanding the menu on render and some custom styling. The `expanded` attribute does work but the menu is not positioned correctly on the screen. I had to call `expanded = true` explicitly to have the menu expanded and positioned properly.

**Comment by pei-jung** [comment]
Hi Chema, did you get a chance to look at this?

Thanks!

**Comment by jbalsas** [comment]
Hey @pei-jung, sorry I haven't been able to look at this yet. I'll make sure to go over it first thing tomorrow!

**Comment by jbalsas** [comment]
Hey @pei-jung, sorry, it's very hard to wrap my head around how this is structured. It doesn't make sense to me, but surely you might see it differently.

What I'd expect is:
- A service that returns `items` json (**not necessary unless this is the source of the performance regression**)
- A simple jsp markup that renders the `dropdown` label
- A single instantiation of `ClayDropdown`

What doesn't feel right:
- A jsp that renders html and js and needs to be 'evaled'
- A jsp that renders basic markup and then on fetch brings in additional html when it could be simply fetching a json collection and instantiating the items.

Could you please take a look at https://github.com/slnn/liferay-portal/pull/360/commits/7790e364b33836e38cf2784b211471e14e680f3c#diff-c6cbd4d8bcd50c04e1bdf4cde9f57282 where I did this as a POC to prove this can be used without a performance impact?

/cc @drewbrokke 

Thanks!

**Comment by pei-jung** [comment]
Hi Chema,

Thanks for looking into this. The rendering of the items is the cause of the performance regression when user signs in, which is why I tried to only load the dropdown after click. Doing so helps prevent an impact on the page load time. We don't have a web service available that returns the items as json, and I'm not sure how to create one. So what I did instead is a simplified version of what you suggested 

1.  Create a simple jsp markup that renders the dropdown label
2. Instantiate the `ClayDropdown` when the label is clicked

I sent the changes to Lily for performance testing in https://github.com/slnn/liferay-portal/pull/364. If it passes the performance tests, I think we're good to go. Thanks again for your help!

PS. the dropdown items are registered as OSGi services. You'll need to redeploy all the modules that I touched to get the items to show up in the dropdown. There's quite a few: `product-navigation-user-personal-menu-api`, `product-navigation-user-personal-bar-web`, `users-admin-web`, `notifications-web`, `sharing-web`, `portal-workflow-web`, etc.

**Pull request closed without merging**